src/main/scala/org/apache/spark/sql/pulsar/PulsarHelper.scala

```diff
@@ -394,106 +394,71 @@ private[pulsar] case class PulsarHelper(
-        } else {
-          assert(time > 0, s"time less than 0: $time")
-          val reader = client
-            .newReader()
-            .subscriptionRolePrefix(driverGroupIdPrefix)
-            .topic(tp)
-            .startMessageId(MessageId.earliest)
-            .startMessageIdInclusive()
-            .create()
-
-          var earliestMessage: Message[Array[Byte]] = null
-          earliestMessage = reader.readNext(pollTimeoutMs, TimeUnit.MILLISECONDS)
-          if (earliestMessage == null) {
-            UserProvidedMessageId(MessageId.earliest)
-          } else {
-            val earliestId = earliestMessage.getMessageId
-
-            reader.seek(time)
-            var msg: Message[Array[Byte]] = null
-            msg = reader.readNext(pollTimeoutMs, TimeUnit.MILLISECONDS)
-            reader.close()
-            if (msg == null) {
-              UserProvidedMessageId(MessageId.earliest)
-            } else {
-              if (msg.getMessageId == earliestId) {
-                UserProvidedMessageId(MessageId.earliest)
-              } else {
-                // intentionally leave this id from UserProvided since it's the last id
-                // less than time, need to skip this one at the beginning
-                PulsarSourceUtils.mid2Impl(msg.getMessageId)
-              }
-            }
-          }
-        }
+        case _: MessageId =>
+          getUserProvidedMessageId(tp, None, Some(off))
+        case _ =>
+          throw new IllegalArgumentException(s"not supported offset type: $off")
+      }
       (tp, actualOffset)
     }
   }
 
   private def fetchCurrentOffsets(
-      offset: SpecificPulsarOffset,
-      poolTimeoutMs: Int,
+      time: SpecificPulsarTime,
       reportDataLoss: String => Unit): Map[String, MessageId] = {
 
-    offset.topicOffsets.map { case (tp, off) =>
-      val actualOffset = fetchOffsetForTopic(poolTimeoutMs, reportDataLoss, tp, off)
+    time.topicTimes.map { case (tp, time) =>
+      val actualOffset = time match {
+        case PulsarProvider.EARLIEST_TIME =>
+          UserProvidedMessageId(MessageId.earliest)
+        case PulsarProvider.LATEST_TIME =>
+          UserProvidedMessageId(PulsarSourceUtils.seekableLatestMid(getLastMessageId(tp)))
+        case _ =>
+          assert(time > 0, s"time less than 0: $time")
+          getUserProvidedMessageId(tp, Some(time), None)
+      }
       (tp, actualOffset)
     }
   }
 
-  @tailrec
-  private def fetchOffsetForTopic(
-      poolTimeoutMs: Int,
-      reportDataLoss: String => Unit,
-      tp: String,
-      off: MessageId): MessageId = {
-    off match {
-      case UserProvidedMessageId(mid) if mid == MessageId.earliest =>
-        UserProvidedMessageId(mid)
-      case UserProvidedMessageId(mid) if mid == MessageId.latest =>
-        UserProvidedMessageId(mid)
-      case UserProvidedMessageId(mid) =>
-        fetchOffsetForTopic(poolTimeoutMs, reportDataLoss, tp, mid)
```

Comment:
The intention of this code segment is not clear to me.
We can handle them later if needed.